packages/react/src/internal/Selection.tsx

```diff
@@ -0,0 +1,242 @@
+/**
+ * Copyright IBM Corp. 2016, 2025
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, { useCallback, useEffect, useState, useRef } from 'react';
+import isEqual from 'react-fast-compare';
+
+const callOnChangeHandler = <ItemType,>({
+  isControlled,
+  isMounted,
+  onChangeHandlerControlled,
+  onChangeHandlerUncontrolled,
+  selectedItems,
+}: {
+  isControlled: boolean;
+  isMounted: boolean;
+  onChangeHandlerControlled?: (data: { selectedItems: ItemType[] }) => void;
+  onChangeHandlerUncontrolled: (selectedItems: ItemType[]) => void;
+  selectedItems: ItemType[];
+}) => {
+  if (isControlled) {
+    if (isMounted && onChangeHandlerControlled) {
+      onChangeHandlerControlled({ selectedItems });
+    }
+  } else {
+    onChangeHandlerUncontrolled(selectedItems);
+  }
+};
+
+interface UseSelectionProps<ItemType> {
+  disabled?: boolean;
+  filteredItems?: ItemType[];
+  initialSelectedItems?: ItemType[];
+  onChange?: (data: { selectedItems: ItemType[] }) => void;
+  selectAll?: boolean;
+  selectedItems?: ItemType[];
+}
+
+export const useSelection = <ItemType,>({
+  disabled,
+  onChange,
+  initialSelectedItems = [],
+  selectedItems: controlledItems,
+  selectAll = false,
+  filteredItems = [],
+}: UseSelectionProps<ItemType>) => {
+  const isMounted = useRef(false);
+  const savedOnChange = useRef(onChange);
+  const [uncontrolledItems, setUncontrolledItems] =
+    useState(initialSelectedItems);
+  const isControlled = !!controlledItems;
+  const selectedItems = isControlled ? controlledItems! : uncontrolledItems;
+  const onItemChange = useCallback(
+    (item: ItemType) => {
+      if (disabled) return;
+
+      // Assert special properties (e.g., `disabled`, `isSelectAll`, etc.) are
+      // `any` since those properties aren’t part of the generic type.
+      const allSelectableItems = filteredItems.filter(
+        (item) => !(item as any)?.disabled
+      );
+      const disabledItemCount = filteredItems.filter(
+        (item) => (item as any)?.disabled
+      ).length;
+
+      let newSelectedItems: ItemType[];
+
+      // deselect all on click to All/indeterminate option
+      if ((item as any)?.isSelectAll && selectedItems.length > 0) {
+        newSelectedItems = [];
+      }
+      // select all options
+      else if ((item as any)?.isSelectAll && selectedItems.length === 0) {
+        newSelectedItems = allSelectableItems;
+      } else {
+        let selectedIndex: number | undefined;
+        selectedItems.forEach((selectedItem, index) => {
+          if (isEqual(selectedItem, item)) {
+            // TODO: Should this functionality be changed so that the loop stops
+            // iterating after finding a match? If not, should the loop be
+            // iterated in reverse? Either way, seems like there's a more
+            // performant way to do this.
+            selectedIndex = index;
+          }
```

Comment:
Yeah I'm not sure if it _needs_ to be the last selected index, but that's what the result of this ends up being. Feels like it could instead be this?
```suggestion
        selectedIndex = selectedItems.findLastIndex(selectedItem => isEqual(selectedItem, item));
```

Reply:
Yea, a solution with `findLastIndex` worked locally. I was able to convert the variable to a `const` which is better too.